### DISCUSSION TITLE
[FeatureRequest] Elasticsearch: allow to remove ".keyword" suffix for grouping fields

### DISCUSSION THREAD

**Discussion by M0rdecay** [grafana/grafana#39723] [Alerting]
**What would you like to be added**:
A setting that allows change a names of fields in alerts
Unfortunately, transformations are not available for panels with alerts:
![image](https://user-images.githubusercontent.com/50422107/115836980-c838b380-a420-11eb-8d7e-c269f9313a31.png)

**Why is this needed**:
By default, Elasticsearch allows index a field twice - for searching and for grouping.
A field indexed for grouping gets `.keyword` suffix:
![image](https://user-images.githubusercontent.com/50422107/115836582-506a8900-a420-11eb-90df-a9c03e7a5f49.png)

This is how a field is displayed in alert:
![image](https://user-images.githubusercontent.com/50422107/115836740-8576db80-a420-11eb-89c9-31b9f330f4c2.png)

This is very inconvenient when you need to match events by the same fields from different databases:
![image](https://user-images.githubusercontent.com/50422107/115837221-ffa76000-a420-11eb-8f92-ab7717a6a72a.png)

**Comment by hugohaggmark** [comment]
Passing this over to @grafana/observability-squad

**Comment by ivanahuckova** [comment]
In Grafana 8+ we are moving away from having alerting tied to graph panel. Alerting queries are created in separate UI and therefore transformations that are used in the panel won't affect them.

@Elfo404 @nathanrodman @peterholmberg Do you think that this feature request should be solved on data source level (we would implement this only in ElasticSearch) or if other data sources would benefit from this as well and this should be a new alerting feature request.

**Comment by Elfo404** [comment]
@ivanahuckova while I don't see this being difficult to implement on a DS level, I'd say it would be better suited for a generic alerting "renaming" feature.

looping in @grafana/alerting-squad

**Comment by ifrost** [comment]
@grafana/alerting-squad what do you think about it?

**Comment by yuri-tceretian** [comment]
> By default, Elasticsearch allows index a field twice - for searching and for grouping.
A field indexed for grouping gets .keyword suffix

<details><summary>My 2 cents about Elasticsearch</summary>

`keyword` is just a name of the nested field. This comes from the default template - every time Elasticsearch sees a new field in an index it creates a dynamic mapping unless it is not explicitly disabled, in which case, it will fail the request. The default mapping for a string field is like the following:
```
{
  ....
 "hostname": {
    "fields": {
      "keyword": {
        "ignore_above": 256,
        "type": "keyword"
      }
    },
    "type": "text"
  },
...
}
```
As you can see there is a field `hostname` and a field `hostname.keyword` which in fact are two different fields and are stored in two different Lucene indices. 
In Elasticsearch the nested fields are very common, the Elastic Common Schema is a good example. Therefore, the field name "hostname.keyword" is no different than "docker.healthcheck.event.end_date" from a querying perspective.

The default mapping for text fields is rarely chosen consciously. Usually, users define either of those while they define the index mapping or a template. 

From what I can see in your screenshot, you group by hostname and application name. I think that for both fields the text type of the field is not necessary, and getting rid of it can reduce index size as well as improve ingestion rate. You can still search by keyword field. The search will not be as flexible as against a text field but in my experience, it should be enough.

</details>

If I understand the request correctly, Grafana should provide the ability to map labels. This feature can be useful in many ways. I believe that it should be possible to implement it in Grafana hosted alerts as we have more control over the configuration. However, it is still not a trivial feature to support. It will require a serious amount of work in the backend as well as UI work. Therefore, unfortunately, it is not possible to do so in the near future as there are plenty of more high-priority tasks. We will put it in our backlog. 

cc @armandgrillet, @gotjosh we discussed that during our weekly meeting, and agreed that this is a nice proposal. What's the next in our process?

**Comment by armandgrillet** [comment]
The issue https://github.com/grafana/grafana/issues/42023 has been created following this discussion, it is in the Alerting backlog and needs more investigation: https://github.com/orgs/grafana/projects/52. Please continue the discussion in the issue directly, the goal with this issue is to not forget about it.

**Comment by gabor** [comment]
it seems the issue will be handled&solved in https://github.com/grafana/grafana/issues/42023 , so i think we can close this discussion.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
